keyring.php

```diff
@@ -129,6 +129,18 @@ function() {
 	 */
 	static function request_handlers() {
 		global $current_user;
+		
+		// Some services do not return the redirect_uri params, in this case these are packed into the state param.
+		// If so, they are unpacked again here and merged back into the request globals.
+		if ( ! empty( $_REQUEST['state'] ) ) {
+			$unpacked_state = unserialize( base64_decode( $_REQUEST['state'] ) );
+			if ( ! empty( $unpacked_state['action'] )) {
+				foreach ( $unpacked_state as $key => $value) {
```

Comment:
On thinking about it we should probably just add the specific params we want back into the request params, this was easy for testing purposes, but probably safer to not just add everything that could came back in the serialised state into the request params indiscriminately.

Reply:
Yes, I think picking out what we need would probably be best. I also see that the [recommendation is to encrypt/sign the `state` parameter](https://auth0.com/docs/protocols/oauth2/redirect-users#alternate-method) when this technique is used, to prevent tampering.

Should we maybe use the client secret to encrypt the base 64 encoded string?